File: fullstack/backend/app/api/routes/chatbot.py
# Author: Anh Ho
# Function: API endpoints for the AI Chatbot.

from fastapi import APIRouter, HTTPException
import logging

logger = logging.getLogger(__name__)

from app.api.deps import CurrentUser, SessionDep
from app.schemas.chatbot import ChatRequest, ChatResponse
from app.services.chatbot.orchestrator import handle_chat

router = APIRouter(prefix="/chatbot", tags=["chatbot"])


@router.post("/chat", response_model=ChatResponse)
async def chatbot_chat(payload: ChatRequest, session: SessionDep, current_user: CurrentUser):
    # Attempts to call the handle chat function to see if the chatbot is functioning.
    try:
        return await handle_chat(
            message=payload.message,
            project_id=payload.project_id,
            session=session,
            current_user=current_user,
        )

    # Handles error if the chatbot failed.
    except Exception as exc:
        logger.exception("Chatbot request failed")

        return ChatResponse(
            response="Sorry, I could not process your request right now.",
            command_used=None,
            command_data=None,
        )

File: fullstack/frontend/src/api/chatbot.ts
import { api } from "./client"

export type ChatbotRequest = {
  message: string
  project_id?: string | number
}

export type CommandUsed = {
  command: string
  arguments: Record<string, unknown>
}

export type ChatbotResponse = {
  response: string
  command_used?: string
  command_data?: unknown
  commands_used?: CommandUsed[]
  command_results?: unknown[]
}

export const chatbotApi = {
  sendMessage: (payload: ChatbotRequest) =>
    api.post<ChatbotResponse>("/chatbot/chat", payload).then((res) => res.data),
}